### DISCUSSION TITLE
Redirect mechanism during IKE_SA_INIT

### DISCUSSION THREAD

**Discussion by zekena2** [strongswan/strongswan#2227] [Q&A]
The [RFC5685](https://datatracker.ietf.org/doc/html/rfc5685) states that "The redirect can be done during the IKE_SA_INIT or the IKE_AUTH exchange" How can I configure that with strongswan? VICI supports doing the redirect only on established SAs.

**Comment by tobiasbrunner** [comment]
To redirect during these exchanges you need to write a plugin that implements the `redirect_provider_t` interface. There is an [example out-of-tree plugin](https://github.com/tobiasbrunner/strongswan-redirect-plugin).

**Reply by zekena2** [reply]
I got this plugin by compiling but how do I configure it? how can I configure the gateway and the IKE name?

**Reply by tobiasbrunner** [reply]
The plugin does nothing out of the box and has no configuration option. You need to change the code so it does whatever you need it to do.

**Reply by zekena2** [reply]
I think this RFC should be "under development" in this list https://docs.strongswan.org/docs/5.9/features/ietf.html as it's not usable outside of using swanctl on established SAs. Comparing strongswan to libreswan here libreswan has better support out of the box for this feature.

**Reply by tobiasbrunner** [reply]
It's perfectly usable, with even more flexibility than is offered by other implementations. What use case(s) for the redirect mechanism are you considering anyway?

**Reply by zekena2** [reply]
I'm considering it for load balancing in order to have a highly available system as well. for example in libreswan it allows to specify a list of servers to redirect to on the IKE_SA_INIT step and does some kind of round robin load balancing between these servers. Additionally I'd like to update this list without restarting the server. This could be enough but maybe I'd like to have this redirect server does a health check on these servers as well and only serve the healthy ones but I could do that part myself as well.

**Reply by tobiasbrunner** [reply]
> I'm considering it for load balancing in order to have a highly available system as well. for example in libreswan it allows to specify a list of servers to redirect to on the IKE_SA_INIT step and does some kind of round robin load balancing between these servers.

Why use this mechanism for that (which not all clients will support)? Why not just do it via DNS?

> Additionally I'd like to update this list without restarting the server.

Which you can easily do in a plugin via the optional [`reload()` method](https://github.com/strongswan/strongswan/blob/287ef047a9f4213f377302e6cb70d52a6427878a/src/libstrongswan/plugins/plugin.h#L53-L58), which is called after strongswan.conf has been reloaded (triggered by `swanctl --reload-settings` or a `SIGHUP` to the daemon).


> but maybe I'd like to have this redirect server does a health check on these servers as well and only serve the healthy ones

Exactly the reason why we provide this plugin architecture. This offers you the most flexibility to do the redirection based on whatever information you have available in the way you see fit.

**Reply by zekena2** [reply]
Clients are controlled by us they'll use the latest version of strongswan on linux. DNS is less reliable in our setup compared to redirect. 

I will write code anyway that's true and I don't mind writing a plugin unless it's in C (My C experience is very small) but if I have to write code that means the feature isn't supported, right?

**Reply by tobiasbrunner** [reply]
> I will write code anyway that's true and I don't mind writing a plugin unless it's in C (My C experience is very small) 

Depending on what you want to do, there is not that much code to be written. You can also use the data structures and utilities provided by libstrongswan (see [the apidoc](https://www.strongswan.org/apidoc/group__libstrongswan.html)).

You could also call a script or some other program (via [process abstraction layer](https://www.strongswan.org/apidoc/group__process.html)), query a database (there is an [abstraction layer](https://www.strongswan.org/apidoc/group__database__t.html) implemented for MySQL/MariaDB and SQLite), or communicate with an external process over a socket interface (via [stream abstraction](https://www.strongswan.org/apidoc/group__database__t.html)).

> if I have to write code that means the feature isn't supported, right?

The plugin decides when which client is redirected where. That's all outside the scope of the RFC.

**Reply by zekena2** [reply]
I'll see what I can do. Thanks for your help and links provided.
